File: desktop/macos/Sources/VoiceStickApp/BleProtocol.swift
import Foundation

struct AudioFrame {
    let sessionID: UInt32
    let seq: UInt32
    let flags: UInt8
    let payload: Data

    var isStart: Bool { flags & 0x01 != 0 }
    var isEnd: Bool { flags & 0x02 != 0 }
}

struct AudioBatch {
    let transportVersion: UInt8
    let sessionID: UInt32
    let firstSeq: UInt32
    let packetCount: UInt8
    let flags: UInt8
    let frames: [AudioFrame]

    var acknowledgedNextSeq: UInt32 {
        firstSeq &+ UInt32(packetCount)
    }
}

struct StateEvent: Decodable {
    let event: String
    let button: String?
    let sessionID: UInt32?
    let durationMs: UInt32?
    let hardware: String?
    let firmwareVersion: String?
    let buttons: [String]?
    let uiStates: [String]?

    enum CodingKeys: String, CodingKey {
        case event
        case button
        case sessionID = "session_id"
        case durationMs = "duration_ms"
        case hardware
        case firmwareVersion = "firmware_version"
        case buttons
        case uiStates = "ui_states"
    }
}

struct FirmwareOTAStateEvent: Decodable {
    let event: String
    let transferID: UInt32?
    let written: UInt32?
    let size: UInt32?
    let code: String?
    let espErr: Int?
    let rebootMs: Int?

    enum CodingKeys: String, CodingKey {
        case event
        case transferID = "transfer_id"
        case written
        case size
        case code
        case espErr = "esp_err"
        case rebootMs = "reboot_ms"
    }
}

enum BleProtocol {
    static let serviceUUID = "8F2F0B84-6E6F-4B23-88F7-3A3CEAFC5100"
    static let audioUUID = "8F2F0B84-6E6F-4B23-88F7-3A3CEAFC5101"
    static let stateUUID = "8F2F0B84-6E6F-4B23-88F7-3A3CEAFC5102"
    static let controlUUID = "8F2F0B84-6E6F-4B23-88F7-3A3CEAFC5103"
    static let otaRXUUID = "8F2F0B84-6E6F-4B23-88F7-3A3CEAFC5104"
    static let otaStateUUID = "8F2F0B84-6E6F-4B23-88F7-3A3CEAFC5105"

    static let otaTypeBegin: UInt8 = 0x20
    static let otaTypeData: UInt8 = 0x21
    static let otaTypeEnd: UInt8 = 0x22
    static let otaTypeAbort: UInt8 = 0x23
    static let otaTypeState: UInt8 = 0x30
    static let audioType: UInt8 = 0x01
    static let audioAckType: UInt8 = 0x02

    static func parseAudioFrame(_ data: Data) -> AudioFrame? {
        guard let batch = parseAudioBatch(data), batch.frames.count == 1 else { return nil }
        return batch.frames[0]
    }

    static func parseAudioBatch(_ data: Data) -> AudioBatch? {
        guard data.count >= 16 else { return nil }
        let version = data[0]
        guard (version == 1 || version == 2), data[1] == audioType else { return nil }

        let headerLength = UInt16(littleEndianBytes: data[2..<4])
        guard headerLength == 16, data.count >= Int(headerLength) else { return nil }

        let sessionID = UInt32(littleEndianBytes: data[4..<8])
        let seq = UInt32(littleEndianBytes: data[8..<12])
        let flags = data[12]
        let packetCount = version == 2 ? data[13] : 1
        let payloadLength = Int(UInt16(littleEndianBytes: data[14..<16]))
        guard data.count == 16 + payloadLength else { return nil }

        if version == 1 {
            let legacyPacketCount: UInt8 = payloadLength == 0 && flags & 0x02 != 0 ? 0 : 1
            let frame = AudioFrame(
                sessionID: sessionID,
                seq: seq,
                flags: flags,
                payload: data.subdata(in: 16..<(16 + payloadLength))
            )
            return AudioBatch(
                transportVersion: version,
                sessionID: sessionID,
                firstSeq: seq,
                packetCount: legacyPacketCount,
                flags: flags,
                frames: [frame]
            )
        }

        if packetCount == 0 {
            guard payloadLength == 0, flags & 0x02 != 0 else { return nil }
            let endFrame = AudioFrame(sessionID: sessionID, seq: seq, flags: flags, payload: Data())
            return AudioBatch(
                transportVersion: version,
                sessionID: sessionID,
                firstSeq: seq,
                packetCount: 0,
                flags: flags,
                frames: [endFrame]
            )
        }

        var cursor = 16
        var frames: [AudioFrame] = []
        frames.reserveCapacity(Int(packetCount))
        for index in 0..<Int(packetCount) {
            guard cursor < data.count else { return nil }
            let packetLength = Int(data[cursor])
            cursor += 1
            guard packetLength > 0, cursor + packetLength <= data.count else { return nil }

            var frameFlags: UInt8 = 0
            if index == 0 { frameFlags |= flags & 0x01 }
            if index == Int(packetCount) - 1 { frameFlags |= flags & 0x02 }
            frames.append(AudioFrame(
                sessionID: sessionID,
                seq: seq &+ UInt32(index),
                flags: frameFlags,
                payload: data.subdata(in: cursor..<(cursor + packetLength))
            ))
            cursor += packetLength
        }
        guard cursor == data.count else { return nil }
        return AudioBatch(
            transportVersion: version,
            sessionID: sessionID,
            firstSeq: seq,
            packetCount: packetCount,
            flags: flags,
            frames: frames
        )
    }

    static func parseStateEvent(_ data: Data) -> StateEvent? {
        guard data.count >= 4, data[0] == 1, data[1] == 0x10 else { return nil }
        let payloadLength = Int(UInt16(littleEndianBytes: data[2..<4]))
        guard data.count >= 4 + payloadLength else { return nil }
        let payload = data.subdata(in: 4..<(4 + payloadLength))
        return try? JSONDecoder().decode(StateEvent.self, from: payload)
    }

    static func parseFirmwareOTAStateEvent(_ data: Data) -> FirmwareOTAStateEvent? {
        guard data.count >= 4, data[0] == 1, data[1] == otaTypeState else { return nil }
        let payloadLength = Int(UInt16(littleEndianBytes: data[2..<4]))
        guard data.count >= 4 + payloadLength else { return nil }
        let payload = data.subdata(in: 4..<(4 + payloadLength))
        return try? JSONDecoder().decode(FirmwareOTAStateEvent.self, from: payload)
    }

    static func uiStatePayload(state: String, text: String) -> Data {
        let payload = [
            "event": "ui_state",
            "state": state,
            "text": text
        ]
        return (try? JSONSerialization.data(withJSONObject: payload)) ?? Data()
    }

    static func interactionModePayload(_ mode: String) -> Data {
        let payload = [
            "event": "interaction_mode",
            "mode": mode
        ]
        return (try? JSONSerialization.data(withJSONObject: payload)) ?? Data()
    }

    static func audioTransportPayload(version: UInt8 = 2) -> Data {
        let payload: [String: Any] = [
            "event": "audio_transport",
            "version": version,
            "profile": "low_latency"
        ]
        return (try? JSONSerialization.data(withJSONObject: payload)) ?? Data()
    }

    static func audioAckPayload(sessionID: UInt32, nextSeq: UInt32) -> Data {
        var data = Data([2, audioAckType, 12, 0])
        data.appendLittleEndian(sessionID)
        data.appendLittleEndian(nextSeq)
        return data
    }

    static func otaBeginPayload(imageSize: UInt32, transferID: UInt32) -> Data {
        var data = Data([1, otaTypeBegin, 12, 0])
        data.appendLittleEndian(imageSize)
        data.appendLittleEndian(transferID)
        return data
    }

    static func otaDataPayload(transferID: UInt32, offset: UInt32, chunk: Data) -> Data {
        var data = Data([1, otaTypeData, 12, 0])
        data.appendLittleEndian(transferID)
        data.appendLittleEndian(offset)
        data.append(chunk)
        return data
    }

    static func otaEndPayload(transferID: UInt32, imageSize: UInt32) -> Data {
        var data = Data([1, otaTypeEnd, 12, 0])
        data.appendLittleEndian(transferID)
        data.appendLittleEndian(imageSize)
        return data
    }

    static func otaAbortPayload(transferID: UInt32) -> Data {
        var data = Data([1, otaTypeAbort, 8, 0])
        data.appendLittleEndian(transferID)
        return data
    }
}

private extension UInt16 {
    init(littleEndianBytes bytes: Data.SubSequence) {
        self = bytes.enumerated().reduce(0) { $0 | UInt16($1.element) << UInt16($1.offset * 8) }
    }
}

private extension UInt32 {
    init(littleEndianBytes bytes: Data.SubSequence) {
        self = bytes.enumerated().reduce(0) { $0 | UInt32($1.element) << UInt32($1.offset * 8) }
    }
}

private extension Data {
    mutating func appendLittleEndian(_ value: UInt32) {
        append(UInt8(value & 0xff))
        append(UInt8((value >> 8) & 0xff))
        append(UInt8((value >> 16) & 0xff))
        append(UInt8((value >> 24) & 0xff))
    }
}

File: desktop/windows/src/ble_protocol.cc
#include "ble_protocol.h"

#include <algorithm>
#include <cctype>
#include <charconv>

namespace voicestick {

namespace {

std::string TrimCopy(std::string_view text) {
    auto begin = text.begin();
    auto end = text.end();
    while (begin != end && std::isspace(static_cast<unsigned char>(*begin))) ++begin;
    while (begin != end && std::isspace(static_cast<unsigned char>(*(end - 1)))) --end;
    return std::string(begin, end);
}

std::string Uppercase(std::string value) {
    std::transform(value.begin(), value.end(), value.begin(), [](unsigned char c) {
        return static_cast<char>(std::toupper(c));
    });
    return value;
}

bool IsHex4(std::string_view text) {
    return text.size() == 4 && std::all_of(text.begin(), text.end(), [](unsigned char c) {
        return std::isxdigit(c) != 0;
    });
}

std::string JsonStringValue(std::string_view json, std::string_view key) {
    const std::string needle = "\"" + std::string(key) + "\"";
    auto key_pos = json.find(needle);
    if (key_pos == std::string_view::npos) return {};
    auto colon = json.find(':', key_pos + needle.size());
    if (colon == std::string_view::npos) return {};
    auto first_quote = json.find('"', colon + 1);
    if (first_quote == std::string_view::npos) return {};
    std::string out;
    bool escaped = false;
    for (auto i = first_quote + 1; i < json.size(); ++i) {
        char ch = json[i];
        if (escaped) {
            switch (ch) {
            case '"': out.push_back('"'); break;
            case '\\': out.push_back('\\'); break;
            case '/': out.push_back('/'); break;
            case 'b': out.push_back('\b'); break;
            case 'f': out.push_back('\f'); break;
            case 'n': out.push_back('\n'); break;
            case 'r': out.push_back('\r'); break;
            case 't': out.push_back('\t'); break;
            default: out.push_back(ch); break;
            }
            escaped = false;
        } else if (ch == '\\') {
            escaped = true;
        } else if (ch == '"') {
            return out;
        } else {
            out.push_back(ch);
        }
    }
    return {};
}

std::optional<std::uint32_t> JsonU32Value(std::string_view json, std::string_view key) {
    const std::string needle = "\"" + std::string(key) + "\"";
    auto key_pos = json.find(needle);
    if (key_pos == std::string_view::npos) return std::nullopt;
    auto colon = json.find(':', key_pos + needle.size());
    if (colon == std::string_view::npos) return std::nullopt;
    auto begin = colon + 1;
    while (begin < json.size() && std::isspace(static_cast<unsigned char>(json[begin]))) ++begin;
    auto end = begin;
    while (end < json.size() && std::isdigit(static_cast<unsigned char>(json[end]))) ++end;
    if (begin == end) return std::nullopt;
    std::uint32_t value = 0;
    auto result = std::from_chars(json.data() + begin, json.data() + end, value);
    if (result.ec != std::errc()) return std::nullopt;
    return value;
}

std::string JsonEscape(std::string_view text) {
    std::string out;
    out.reserve(text.size() + 8);
    for (char ch : text) {
        switch (ch) {
        case '\\': out += "\\\\"; break;
        case '"': out += "\\\""; break;
        case '\n': out += "\\n"; break;
        case '\r': out += "\\r"; break;
        case '\t': out += "\\t"; break;
        default: out.push_back(ch); break;
        }
    }
    return out;
}

} // namespace

std::optional<AudioFrame> BleProtocol::ParseAudioFrame(std::span<const std::uint8_t> data) {
    if (data.size() < 16 || data[0] != 1 || data[1] != 0x01) return std::nullopt;
    const auto header_len = ReadLe16(data.subspan(2, 2));
    if (header_len != 16 || data.size() < header_len) return std::nullopt;
    const auto payload_len = ReadLe16(data.subspan(14, 2));
    if (data.size() < 16u + payload_len) return std::nullopt;
    AudioFrame frame;
    frame.session_id = ReadLe32(data.subspan(4, 4));
    frame.seq = ReadLe32(data.subspan(8, 4));
    frame.flags = data[12];
    frame.payload.assign(data.begin() + 16, data.begin() + 16 + payload_len);
    return frame;
}

std::optional<StateEvent> BleProtocol::ParseStateEvent(std::span<const std::uint8_t> data) {
    if (data.size() < 4 || data[0] != 1 || data[1] != 0x10) return std::nullopt;
    const auto payload_len = ReadLe16(data.subspan(2, 2));
    if (data.size() < 4u + payload_len) return std::nullopt;
    const auto json = Utf8FromBytes(data.subspan(4, payload_len));
    StateEvent event;
    event.event = JsonStringValue(json, "event");
    if (event.event.empty()) return std::nullopt;
    event.button = JsonStringValue(json, "button");
    event.session_id = JsonU32Value(json, "session_id");
    event.duration_ms = JsonU32Value(json, "duration_ms");
    event.hardware = JsonStringValue(json, "hardware");
    event.firmware_version = JsonStringValue(json, "firmware_version");
    return event;
}

std::optional<FirmwareOtaStateEvent> BleProtocol::ParseFirmwareOtaStateEvent(std::span<const std::uint8_t> data) {
    if (data.size() < 4 || data[0] != 1 || data[1] != ota_type_state) return std::nullopt;
    const auto payload_len = ReadLe16(data.subspan(2, 2));
    if (data.size() < 4u + payload_len) return std::nullopt;
    const auto json = Utf8FromBytes(data.subspan(4, payload_len));
    FirmwareOtaStateEvent event;
    event.event = JsonStringValue(json, "event");
    if (event.event.empty()) return std::nullopt;
    event.transfer_id = JsonU32Value(json, "transfer_id");
    event.written = JsonU32Value(json, "written");
    event.size = JsonU32Value(json, "size");
    event.code = JsonStringValue(json, "code");
    event.reboot_ms = JsonU32Value(json, "reboot_ms");
    return event;
}

ByteVector BleProtocol::UiStatePayload(std::string_view state, std::string_view text) {
    const auto json = std::string("{\"event\":\"ui_state\",\"state\":\"") +
                      JsonEscape(state) + "\",\"text\":\"" + JsonEscape(text) + "\"}";
    return ByteVector(json.begin(), json.end());
}

ByteVector BleProtocol::InteractionModePayload(std::string_view mode) {
    const auto json = std::string("{\"event\":\"interaction_mode\",\"mode\":\"") +
                      JsonEscape(mode) + "\"}";
    return ByteVector(json.begin(), json.end());
}

ByteVector BleProtocol::OtaBeginPayload(std::uint32_t image_size, std::uint32_t transfer_id) {
    ByteVector data = {1, ota_type_begin, 12, 0};
    AppendLe32(data, image_size);
    AppendLe32(data, transfer_id);
    return data;
}

ByteVector BleProtocol::OtaDataPayload(std::uint32_t transfer_id,
                                       std::uint32_t offset,
                                       std::span<const std::uint8_t> chunk) {
    ByteVector data = {1, ota_type_data, 12, 0};
    AppendLe32(data, transfer_id);
    AppendLe32(data, offset);
    data.insert(data.end(), chunk.begin(), chunk.end());
    return data;
}

ByteVector BleProtocol::OtaEndPayload(std::uint32_t transfer_id, std::uint32_t image_size) {
    ByteVector data = {1, ota_type_end, 12, 0};
    AppendLe32(data, transfer_id);
    AppendLe32(data, image_size);
    return data;
}

ByteVector BleProtocol::OtaAbortPayload(std::uint32_t transfer_id) {
    ByteVector data = {1, ota_type_abort, 8, 0};
    AppendLe32(data, transfer_id);
    return data;
}

std::optional<std::string> BleProtocol::DeviceIdFromName(std::string_view name) {
    auto value = Uppercase(TrimCopy(name));
    if (!value.starts_with("VS-")) return std::nullopt;
    value = value.substr(3, 4);
    if (!IsHex4(value)) return std::nullopt;
    return value;
}

std::string BleProtocol::NormalizeDeviceId(std::string_view text) {
    auto value = Uppercase(TrimCopy(text));
    if (value.starts_with("VS-")) value = value.substr(3);
    value = value.substr(0, std::min<std::size_t>(4, value.size()));
    return IsHex4(value) ? value : std::string();
}

} // namespace voicestick
